docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,20 +2,187 @@
+ |              |    | p2 |      |  p1 |     |   |   |   |A(p2)-= N
+ |2             |    |    |      |     |p1,p2|   |   |   |A(i)+=VP(i)
+ |              | p1 |    |      |     |  p2 |   |   |   |A(p1)-= N
+ |3             |    | p1 |      |     |     |   |   | p2|A(i)+=VP(i)
+ |              |    | p1 |      | p2  |     |   |   |   |A(p2)-= N
+ |4             |    |    |   p1 |     |     |   | p2|   |A(i)+=VP(i)
+ |              |    |    |p1,p2 |     |     |   |   |   |A(p2)-= N
+
+The actual implementation does not create this queue, it only has to store the current accumulated proposer priority. 
+
+### Validator Set Changes
+At each block height the validator set may change. Some of the changes have implications on the proposer selection.
+
+#### Voting Power Change
+Consider again the earlier example and assume that the voting power of p1 is changed to 6:
+
+Validator | p1 | p2 
+----------|--- | ---
+VP        | 4  | 3
+
+Let's also assume that the last round R before this change the proposer priorites were as shown in first row (R-1 end). As it can be seen, the procedure can continue with round 1 at next height without changes as before. 
+
+|Priority Round| -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Comment
+|--------------| ---|--- |------|---  |---  |---|---|---|--------
+| R end        |    | p2 |      |  p1 |     |   |   |   |update VP(p1)
+| 1            |    |    |      |     | p2  |   |   | p1|A(i)+=VP(i)
+|              | p1 |    |      |     | p2  |   |   |   |A(p1)-= N
+
+
+#### Validator Removal
+Consider a new example with set.
+
+Validator | p1 | p2 | p3 |
+--------- |--- |--- |--- |
+VP        | 1  | 2  | 3  |
+
+Let's assume that the last round was R and the proposer priorites were as shown in first row (R end) with their sum being 0. At this point p2 is removed and round 1 of next height runs. At the end of the round (penultimate row) the sum of priorites is -2 (minus the priority of the removed process). 
+ 
+
+|Priority Round |-3 | -2 | -1 | 0  | 1   | 2   | 3 | 4 | 5 | Comment
+|---------------|-- | ---|--- |--- |---  |---  |---|---|---|--------
+| R end         |p3 |    |    |    | p1  | p2  |   |   |   |remove p2
+| 1             |   |    |    | p3 |     | p1  |   |   |   |A(i)+=VP(i)
+|               |   | p1 |    | p3 |     |     |   |   |   |A(p1)-= N
+| *new step*    |   |    | p1 |    | p3  |     |   |   |   |A(i) -= avg
+
+The procedure could continue without modifications. However, it is possible that after a sufficiently large number of modifications in validator set, the priority values would group towards maximum or minimum allowed values.
+For this reasons, the selection procedure adds another step (see last row) that shifts the current priority values left or right such that the average remains 0.
+
+The modified selection algorithm is:
+
+    for each validator i in vset:
+      A(i) += VP(i)
+    prop = max(A)
+    A(prop) -= N
+
+    // normalize - shift with average priority
```

Comment:
Shouldn't we normalise upon validator removal so invariant (sum is 0) holds at the beginning of next round, instead of doing it as part of algorithm (every time)?

Reply:
Yes the code could be changed. Just documenting what goes in 0.29 for the moment. I believe that there are many ideas for changes (just reading the different comments in related issues). Also @Liamsi may have comments.

Reply:
Will change in next release.